torch/jit/_script.py

```diff
@@ -270,6 +269,154 @@ def __getattr__(self, attr):
+            if self._c._has_method("__iadd__"):
+                return self.forward_magic_method("__iadd__", other)
+            else:
+                return self.forward_magic_method("__add__", other)
+
+        def __sub__(self, other):
+            return self.forward_magic_method("__sub__", other)
+
+        def __pow__(self, other):
+            return self.forward_magic_method("__pow__", other)
+
+        def __truediv__(self, other):
+            return self.forward_magic_method("__truediv__", other)
+
+        def __mod__(self, other):
+            return self.forward_magic_method("__mod__", other)
+
+        def __ne__(self, other):  # noqa T484
+            return self.forward_magic_method("__ne__", other)
+
+        def __eq__(self, other):  # noqa T484
+            return self.forward_magic_method("__eq__", other)
+
+        def __lt__(self, other):
+            return self.forward_magic_method("__lt__", other)
+
+        def __gt__(self, other):
+            return self.forward_magic_method("__gt__", other)
+
+        def __le__(self, other):
+            return self.forward_magic_method("__le__", other)
+
+        def __ge__(self, other):
+            return self.forward_magic_method("__ge__", other)
+
+        def __and__(self, other):
+            return self.forward_magic_method("__and__", other)
+
+        def __or__(self, other):
+            return self.forward_magic_method("__or__", other)
+
+        def __xor__(self, other):
+            return self.forward_magic_method("__xor__", other)
+
+        def __getitem__(self, other):
+            return self.forward_magic_method("__getitem__", other)
+
+        def __setitem__(self, idx, val):
+            return self.forward_magic_method("__setitem__", idx, val)
+
+        def __call__(self, val):
+            return self.forward_magic_method("__call__", val)
+
+        def __int__(self):
+            return self.forward_magic_method("__int__")
+
+        def __float__(self):
+            return self.forward_magic_method("__float__")
+
+        def __bool__(self):
+            return self.forward_magic_method("__bool__")
+
+        def __str__(self):
+            return self.forward_magic_method("__str__")
+
+        def __enter__(self):
+            return self.forward_magic_method("__enter__")
+
+        def __exit__(self, type, tb, traceback):
+            return self.forward_magic_method("__exit__", type, tb, traceback)
+
+        # TODO: this is a hack to manually call __init__ inside __setstate__ method. Since unpickling
+        # in python3 doesn't call the costructor deliberately, there is no way we can forward the
+        # __setstate__ method to its' cpp class. Therefore, this reduce method manually calls init inside
+        # __setstate__ so that we can avoid infinite recursion. The solution was suggested here:
+        # https://stackoverflow.com/questions/50308214/python-3-alternatives-for-getinitargs
+        def __reduce__(self):
```

Comment:
this is bit weird.

Reply:
Can you explain the need for this further?

Reply:
Basically when we call **__setstate__** in python, it doesn't actually cal inside **__init__** inside so it infinetely recurses inside **__getattr__** because we don't have the properties inside. Therefore, we can use **__reduce__** to call the constructor so that the props are set.